enable typescript services and sdl

jmreidy opened the pull request:
### What
Enables the generation of `sdl` and `service` files in TypeScript. Uses the `--typescript` boolean flag passed to `generate service` and `generate sdl`. Default remains javascript, but we'll reduce templates to only include TS (and generate the JS via transpiling the TS output).

Implements portions of #523.

### Testing
Added unit tests to cover TS service and SDL files

### Implementation Plan
- [x] Add typescript flag and template to sdl
- [x] Add typescript flag and template to service
- [x] Update sdl tests
- [x] Update service tests
- [x] Add Prisma typescript types to generated TS services
- [x] Ensure sdl and service generation work with TS via `scaffold` command
- [x] Update sdl to use typescript, and generate JS via transpilation
- [x] Update service to use typescript, and generate JS via transpilation
- [x] Add javascript flag to service, default to true
- [x] Add javascript flag to sdl, default to true
- [x] Test all combinations of `destroy` and `generate`
- [x] Enable destroy with TS

cannikin commented:
> ### What
> Enables the generation of `sdl` and `service` files in TypeScript. Uses the `--typescript` boolean flag passed to `generate service` and `generate sdl`. Defaults to `false`
> 
> ### Testing
> Tested locally with redwood-tools. Is there a way of testing via unit tests?

Check out the `__tests__` directory under each generator, there are tests that the correct file is created at the correct path and that the generated content matches a fixture.

jmreidy commented:
@cannikin thanks, I'll mark this as draft until i add tests

peterp commented:
@jmreidy another approach that we've been thinking of is to make the templates typescript by default, but to transpile them to JavaScript and to format them through prettier once they're generated.

The reason for this is because I think having to maintain two separate files for the same template might be a bit cumbersome.

peterp commented:
@jmreidy Yeah, I think that makes sense.

Once we fully support TypeScript I think we'll store the type of project in `redwood.toml` and then default to that in the generator, but allow a flag to override it.

thedavidprice commented:
Excited about this. Thanks @jmreidy 

> How would a consumer enable JS instead of TS? Pass a --typsescript false flag?

Just to be clear, for the near-term any `generate` command should default to JS and require using an option for TS. So in this case, the source will be TS and will by default be transpiled/formatted to JS when generated.

I'm suggesting we maintain JS default across the entire project until JS is no longer the default across the entire project. Does this make sense and do you agree @peterp?

thedavidprice commented:
Ah, also some potential overlap here with scaffold generation that will need to be handled.

@cannikin If the SDL and Service templates are written as TS files, then the `generate scaffold` command will also need to handle them appropriately as well, correct?

cannikin commented:
That's true, the scaffold generator makes a call over to the sdl and service generators' `files()` function to get the list of files to create... so yeah it would need to pass along any new flags that are sent to the scaffold generator!

cannikin commented:
If I can throw in a flag idea, if we're defaulting to typescript I would think just passing along `--javascript` would be enough, rather than `--typescript=false`. I always like "positive" flags to change behavior rather than negative ones, if it makes sense in context.

peterp commented:
> I'm suggesting we maintain JS default across the entire project until JS is no longer the default across the entire project. Does this make sense and do you agree @peterp?

💯, agreed.

jmreidy commented:
@cannikin realizing as well that if we generate TS SDL files, and then run `scaffold`, scaffold would check for the existence of a `js` file and generate one. 

I'm hesitant to add `typescript` to scaffold, since converting web files to use TS is not nearly as straightforward right now (until changes to Router and Cells land etc). 

I'm considering the following:
* enable a `typescript` flag for service and sdl generation, as in this PR
* update scaffold to check for the existence of js OR ts service and SDL files when executing

What do y'all think?

cannikin commented:
Hmm, I don't *love* the idea of only the sdl and services generators supporting TS and not the others...ideally all the generators should offer the same experience. But I guess we have to start somewhere? 

What do you think @thedavidprice and @mojombo? How's the developer experience with only those two generators being typescript enabled? Maybe if you try the `--typescript` flag with any of the other generators you get some console output telling you "Typescript support is coming soon" and we can replace them one at a time?

jmreidy commented:
A couple other options:
* we enable TS for scaffold, but only for non-web files
* we update all generators to support TS, but don't land it until router
support lands

On Fri, May 8, 2020 at 12:27 PM Rob Cameron <notifications@github.com>
wrote:

> Hmm, I don't *love* the idea of only the sdl and services generators
> supporting TS and not the others...ideally all the generators should offer
> the same experience. But I guess we have to start somewhere?
>
> What do you think @thedavidprice <https://github.com/thedavidprice> and
> @mojombo <https://github.com/mojombo>? How's the developer experience
> with only those two generators being typescript enabled? Maybe if you try
> the --typescript flag with any of the other generators you get some
> console output telling you "Typescript support is coming soon" and we can
> replace them one at a time?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redwoodjs/redwood/pull/515#issuecomment-625898010>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAIY3N5LJ5QQBO6YHYDJUDRQQXIXANCNFSM4MZZJ6RQ>
> .
>

thedavidprice commented:
### Handling existing files
>update scaffold to check for the existence of js OR ts service and SDL files when executing

I think this should be true of all the generators, i.e. check for the existence of _both_ TS and JS when executing, and then throw an Error if either exists, regardless of the target being TS or JS. Users can override with `--force`. 

@cannikin do all the generators currently check for existing files before running?

### Scaffold TS support
If it's not going to result in a runtime Error, _and_ if we can communicate the behavior clearly, I'm all for users being able to generate TS files for SDL and Services even though Web will be JS. Most likely this would require a pre-run explanation with user confirmation to proceed. 
> see my comments in the section below explaining `--no-javascript` usage

So maybe:
1. command `yarn rw g scaffold [model] --no-javascript`
2. CLI output `TS support is only available for SDL and Service file types. All Web components will be generated as JS files. Would you like to proceed? y/n`

?

### Generators Requiring Router Support
We can still make the conversion but _only_ allow/enable target=JS (i.e. always transpile when the command is run) until dependencies like the Router are ready. 

Or am I missing something here? 

### Making TS the default Target but enabling JS target as default (for now)
If our goal is to make TS the eventual default for Redwood, I'm wondering how best to navigate user behavior changes during this transition. So imagining a future example:
1. user runs `yarn create redwood-app [dirName] --javascript` --> The default install is TS without the flag. So the user adds the `--javascript` flag to target JS.
2. During installation, the `redwood.toml` project config is set to `target=javascript`
3. Whenever a generator command is run, the `redwood.toml` setting is used as the default
   - a user can override the default with the appropriate `--typescript` or `--javascript` option

So in the future, we'll have a default target to work with. But we don't right now.
- should we add a default now (via Webpack) to set `target=javascript`, which we can then start accessing across all generator commands as we convert?
- and then we can have the generators start 1) reading the App default target and 2) allowing it to be overridden with `--typescript`

I might be overthinking this as we can just add both `--javascript` and `--typescript` options now and set `--javascript` to 'true' as the Yargs default. But then we'll need to come back and add-in the "first check App config target" behavior at a later time.

??

### Organizing Effort with a Tracking issue
Seems like we are going to need some organization here across the entire effort of converting generators to TS. I've created a Tracking Issue #523 

Please take a look and definitely discuss and suggest edits. Once we have consensus on the plan, I'll see if we can get additional community support from [these kind folks as well](https://community.redwoodjs.com/t/team-typescript-for-those-who-want-to-help-add-ts-support/369/19). (@jmreidy would be great to discuss how/if you and Mohsen might want to help with organizing the effort -- no pressure but let us know if interested.)

cannikin commented:
> @cannikin do all the generators currently check for existing files before running?

Yup, they'll stop as soon as they run into a file that exists. You'll need to `--force` to overwrite.

I like the idea of just adding `--javascript` and `--typescript` now and defaulting one now and the other later when we have widespread TS support. But no matter what I heartily veto `--no-javascript` please!

jmreidy commented:
@thedavidprice hats off for fantastic documentation and breakdown in that tracking issue. Thanks so much for the clarifying commentary both there and in this thread.

Two last questions from me:
- do we want to rewrite the yargs generator files themselves as TS? 
- how should we batch these changes? `523_typescript_templates` feature branch?

And yes, happy to help organize the effort and pull in folks from the community, just let me know if you have any specific guidance on how you'd like me to proceed there.

peterp commented:
> do we want to rewrite the yargs generator files themselves as TS?

@jmreidy We're going to modify the architecture of the CLI and generators slightly - I'll involve you in those discussions when they happen, but I don't think we'll get much value out of converting those now.

guledali commented:
Maybe it should look more like this?

![type-arguments](https://user-images.githubusercontent.com/20647282/81516846-a2aeb900-9339-11ea-80d7-67ef32c15c01.png)


If you look at blitz they are using the generated types from prisma client.
https://github.com/blitz-js/blitz/tree/canary/packages/generator/templates/query
https://github.com/blitz-js/blitz/tree/canary/packages/generator/templates/mutation

thedavidprice commented:
@jmreidy Thanks for your patience here as we catch up with your trailblazing! The core team discussed this yesterday. I've updated #523 to address questions here as well as provide a roadmap for proceeding (including some redundancy of previous decisions). 

I'm revising the 523 as we learn more, so please add your comments and questions as needed.

### Questions from previous comments in this PR
>realizing as well that if we generate TS SDL files, and then run scaffold, scaffold would check for the existence of a js file and generate one.
>
>I'm hesitant to add typescript to scaffold, since converting web files to use TS is not nearly as straightforward right now (until changes to Router and Cells land etc).

^^ I attempt to address the first part of this in #523 **tl;dr** generators need to be both TS/JS aware for creating, overwriting, and refusing to overwrite in the case of hybrid scenario.

And for the second part, hybrid generators are fine during this transition. E.g. TS SDL and Services with JS web/ components is ok (for now).
   - however, this is unexpected behavior, so a CLI message+explanation should be given with prompt to "accept and continue"

> do we want to rewrite the yargs generator files themselves as TS?

^^ no, not at this time

> how should we batch these changes? 523_typescript_templates feature branch?

^^ the goal is to transition in steps, meaning each PR will focus on incremental, working changes. This might not always be possible, so we'll adapt as we go. But in this case, it seems right now we can support SDL, Service, and hybrid-Scaffold as an incremental step that does not change existing JS behavior or expectations

jmreidy commented:
@thedavidprice Awesome! This is all the info I need. I'm moving forward with the SDL and service generators, and will start coordinating with the community on related TS issues. I just pushed up unit testing to cover the TS sdl use case, and I'm working on that for services as well.

@guledali great callout, I'm updating services to follow this style, thanks for the poke on it!

jmreidy commented:
@guledali What do you think about how we type the function params?

There's a couple options.

Inline:
```typescript
export const updateUserProfile = ({ id, input }: { id: UserProfileWhereUniqueInput, input: UserProfileUpdateInput }) => {
```

Named args interface:
```typescript
interface UpdateUserProfileArgs {
  id: UserProfileWhereUniqueInput;
  input: UserProfileUpdateInput;
}

export const updateUserProfile = ({ id, input }: UpdateUserProfileArgs) => {
```

thedavidprice commented:
**FYI** 
Windows CI runner is not passing, which is a known, non-blocking issue. #533

guledali commented:
@jmreidy While the inline example is more explicit it also more robust, nice thing about `named` args interface. Is that your eyes can easily scan the type and do a go to definition

This looks clean to me and more readably,

```tsx
interface UpdateUserProfileArgs {
  id: UserProfileWhereUniqueInput;
  input: UserProfileUpdateInput;
}

export const updateUserProfile = ({ id, input }: UpdateUserProfileArgs) => {
``` 

@thedavidprice I want to hear the core team what they have to say?

jmreidy commented on packages/cli/src/commands/generate/service/__tests__/fixtures/multiWordServiceCrud.ts:
```diff
@@ -0,0 +1,45 @@
+  return db.userProfile.findOne({
+    where: { id },
+  })
+}
```
@guledali updated generated files to include types, let me know what you think

guledali replied:
LGTM

jmreidy commented:
I've added tests for SDL and service generation, and updated the generated crud service files to have proper types for function parameters. 

Next I'll need to update the scaffold generator to support typescript, and finally update the generators to use a single TS template and then transpile to JS for default. 

I'll update the PR description to include a checklist reflecting the above.

guledali commented on packages/cli/src/commands/generate/service/templates/service.ts.template:
```diff
@@ -0,0 +1,48 @@
+}<% } %><% if (relations.length) { %>
+
+export const ${singularPascalName} = {<% relations.forEach(relation => { %>
+  ${relation}: (_obj, { root }) => db.${singularCamelName}.findOne({ where: { id: root.id } }).${relation}(),<% }) %>
```
`root` argument can we do something about it?

jmreidy replied:
![image](https://user-images.githubusercontent.com/35949/82123736-e3a73300-9768-11ea-90c6-efaf53fe67dc.png)

Does this look about right?

jmreidy commented on packages/cli/src/lib/index.js:
```diff
@@ -133,7 +134,8 @@ export const generateTemplate = (templateFilename, { name, root, ...rest }) => {
   const parser = {
```
This code wasn't working previously, since `templateFilename` === `template`

jmreidy commented:
@cannikin @peterp @thedavidprice @guledali This should now be ready for review! Thanks for your patience.

thedavidprice commented:
🚀Thanks, trailblazer!

I'll want to take some time with this one since it's the first go at a conversion.  And want to see how things functionally work against #523 Especially things like:
- how transpiled templates compare to original (for Tutorial)
- if we can/should create some additional utils for checking project language and setting defaults --> we might want to include an option for `redwood.toml` setting here at the start
- how things work hybrid js/ts when it comes command being aware of existing in case of overwrite, etc

jmreidy commented on packages/cli/src/commands/generate/sdl/templates/sdl.ts.template:
```diff
@@ -1,3 +1,5 @@
+import gql from 'graphql-tag'
```
It works when building, but VS Code complains about it. Is there a way of configuring VSCode to know to look for the auto-import plugin?

peterp replied:
I think we're going to add ambient declarations in a `node_modules/.redwood` folder that's generated at build-time.

jmreidy replied:
Sounds good, we can strip these when the ambients are added.

jmreidy commented on packages/cli/src/commands/generate/sdl/__tests__/sdl.test.js:
```diff
@@ -1,69 +1,124 @@
 
-  expect(Object.keys(files).length).toEqual(3)
-})
+const itReturnsExactlyThreeFiles = (baseArgs = {}) => {
```
The reason I'm doing this is to avoid duplicating the same set of tests within typescript and javascript description blocks. I'd be happy to employ another solution here. (Alternatively we just duplicate the tests, but that feels slightly brittle to me)

jmreidy replied:
(Although, to argue against myself, I just duped test calls for the scaffold generator in two different `description` blocks...)

jmreidy commented on packages/cli/src/commands/generate/service/__tests__/fixtures/singleWord_belongsTo.ts:
```diff
@@ -0,0 +1,12 @@
+import { UserWhereUniqueInput } from '@prisma/client'
```
Yep exactly. They're not present in the generated JS

jmreidy commented on packages/cli/src/commands/generate/service/__tests__/fixtures/multiWord.test.js:
```diff
@@ -1,4 +1,6 @@
+/*
```
I'm commenting out the import for now, since the generated test doesn't actually do anything with the import. I figure when we start building out the auto-generated test assertions we can remove the comments from the generation. Alternatively I can remove the comments (although a linter will warn about the unused import)

peterp commented on packages/cli/src/lib/index.js:
```diff
@@ -188,6 +190,19 @@ export const prettierOptions = () => {
+export const transformTSToJS = (content) => {
+  const configOptions = prettierOptions()
+  content = babel.transform(content, {
+    plugins: ['@babel/plugin-transform-typescript', 'generator-prettier'],
```
Super interesting, I like it!

jmreidy commented:
@peterp you got it, i will also rebase off master tomorrow so it'll be up to date

peterp commented:
I think it's too late... I can't seem to accept the invite, or figure out how to open a pull-request to your fork; here's my branch:

https://github.com/redwoodjs/redwood/tree/jmreidy-convert_api_to_ts

jmreidy commented:
@peterp I rebased off master and pushed up. The issue was that the scaffold tests expected an object from `builder` instead of the functional pattern. I refactored a touch to extract the default options in scaffold and use those in test (instead of directly using the builder). 

Let me know what else I can add!

Also @cannikin @thedavidprice wanted to give you a chance to review. Thanks all!

peterp commented on packages/cli/src/commands/generate/scaffold/__tests__/scaffold.test.js:
```diff
@@ -22,7 +22,7 @@ describe('in javascript (defualt) mode', () => {
   // SDL
 
-  test('creates an sdl', () => {
+  test.only('creates an sdl', () => {
```
```suggestion
  test('creates an sdl', () => {
```

peterp replied:
Looks like this snuck in!

jmreidy replied:
whoops, fixed!

peterp commented:
I think there may be a problem with the conversion of default to builder, I get the following when trying to run the script:

```terminal
yarn rw g scaffold --help
Generate Pages, SDL, and Services files based on a given DB schema Model. Also
accepts <path/model>.

Positionals:
  model  Model to scaffold. You can also use <path/model> to nest files by type
         at the given path directory (or directories). For example, 'rw g
         scaffold admin/post'.                                        [required]

Options:
  --0
  -1         Generate JavaScript files                 [boolean] [default: true]
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
```

jmreidy commented:
I wasn't destructuring the args correctly, that's fixed now.

Longer term, I wonder if we mock yargs to have true end-to-end tests. I'm a
little worried about skipping the yargs interface entirely to go straight
to the underlying `files` calls.

On Wed, May 27, 2020 at 4:00 PM Peter Pistorius <notifications@github.com>
wrote:

> I think there may be a problem with the conversion of default to builder,
> I get the following when trying to run the script:
>
> yarn rw g scaffold --help
> Generate Pages, SDL, and Services files based on a given DB schema Model. Also
> accepts <path/model>.
>
> Positionals:
>   model  Model to scaffold. You can also use <path/model> to nest files by type
>          at the given path directory (or directories). For example, 'rw g
>          scaffold admin/post'.                                        [required]
>
> Options:
>   --0
>   -1         Generate JavaScript files                 [boolean] [default: true]
>   --help     Show help                                                 [boolean]
>   --version  Show version number                                       [boolean]
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/redwoodjs/redwood/pull/515#issuecomment-634908746>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAIY3I2IP7XQOZ3LJPNDYLRTVWPFANCNFSM4MZZJ6RQ>
> .
>

jmreidy commented:
@peterp Just a heads up that I'm working through testing all combinations of actually running yargs commands locally, and will update this PR when I've confirmed all the smoke tests are working.

jmreidy commented:
Good news! I got it working!
https://www.loom.com/share/9f3542034b3a44cfb974361835c69d46

Bad news - rebasing is a massive pain. Especially after https://github.com/redwoodjs/redwood/pull/611/files, which rewrote all the scaffold templates. 

I'll need to keep working through conflicts, if we can hold on making changes to templates in the meantime, I'd appreciate it...

jmreidy commented on packages/cli/src/commands/destroy/scaffold/__tests__/scaffold.test.js:
```diff
@@ -1,5 +1,21 @@
+import { tasks } from '../scaffold'
+
 jest.mock('fs')
+jest.mock('@babel/core', () => {
```
babel is unhappy when you mock `fs`

peterp replied:
We're thinking of creating a wrapper around `fs` with a few common operations like write, read, unlink in `@redwoodjs/internal` that should be slightly more manageable to mock.

jmreidy commented on packages/cli/src/commands/generate/service/__tests__/service.test.js:
```diff
@@ -1,146 +1,230 @@
 
-  expect(Object.keys(files).length).toEqual(2)
-})
+const itReturnsExactly2Files = (baseArgs) => {
```
noted in a comment elsewhere - extracted tests into named function to reduce duplication across TS and JS versions

jmreidy commented on packages/cli/src/lib/index.js:
```diff
@@ -147,7 +153,32 @@ export const generateTemplate = (templateFilename, { name, root, ...rest }) => {
+}
+
+const getBaseFile = (file) => file.replace(/\.\w*$/, '')
+
```
Instead of adding `typescript` as a flag to `destroy`, it felt more ergonomic to me for it to just destroy whatever js/ts variation is found.

thedavidprice commented:
@jmreidy Woah! Huge congrats to you and yours!! 🎉😍

We can't thank you enough for all the sustained effort you put into this, Justin. It's created so much momentum and paved the way by example. You're going to be missed. But you've made a huge impact here. And we'll be excited to see you again in the future when/if you're available months from now.

One thing that might help --> could you give @peterp (and maybe me) commit access to `jmreidy:convert_api_to_ts`? Not necessary if you don't get around to it.

If for some reason we get absolutely stuck, I’ll reach out to you directly apart from GitHub.

Congrats again!

peterp commented:
Thanks for this @jmreidy!
